clients/cobol-lsp-vscode-extension/src/services/ControlFlowService.ts

```diff
@@ -0,0 +1,311 @@
+/* eslint-disable @typescript-eslint/no-unsafe-enum-comparison */
+/*
+ * Copyright (c) 2025 Broadcom.
+ * The term "Broadcom" refers to Broadcom Inc. and/or its subsidiaries.
+ *
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *   Broadcom, Inc. - initial API and implementation
+ */
+import * as vscode from "vscode";
+import { Program } from "@code4z/analysis/lib/model/cfast";
+import { EngineProcessingResult } from "@code4z/analysis/lib/graphbuilder";
+import { Graph } from "@code4z/analysis/lib/model/Graph";
+import { Worker } from "worker_threads";
+import { join } from "path";
+import {
+  DiagnosticDto,
+  DiagnosticSeverityDto,
+  DiagnosticTagDto,
+} from "@code4z/analysis/lib/model/external";
+import { SettingsService } from "./Settings";
+import { OutputChannelHolder } from "../OutputChannelHolder";
+import { WorkerLoggerMessage, WorkerMessage } from "./worker/messages";
+
+/**
+ * Control Flow Analysis callback
+ */
+export interface ControlFlowAnalysisCallback {
+  (graphs: Graph[]): void;
+}
+
+/**
+ * Adds a Control Flow Analysis callback function, this callback will be fired on analysis finished event
+ * @param documentUri the document uri
+ * @param callback a callback, that will be fired on finish analysis
+ */
+export function addControlFlowAnalysisCallback(
+  documentUri: string,
+  callback: ControlFlowAnalysisCallback,
+): void {
+  ControlFlowAnalysisService.instance().addCallback(documentUri, callback);
+}
+
+/**
+ * Handle AST and starts analysis it
+ */
+export async function controlFlowAstHandler(result: ApiResult) {
+  if (result.documentUri) {
+    await ControlFlowAnalysisService.instance().cancelAnalysis(
+      result.documentUri,
+    );
+    if (result.controlFlowAST.length > 0) {
+      ControlFlowAnalysisService.instance().queueAnalysis(
+        result.controlFlowAST,
+        result.documentUri,
+      );
+    }
+  }
+}
+
+export class ApiResult {
+  public controlFlowAST: Program[] = [];
+  public documentUri: string | undefined;
+}
+
+interface AnalysisServiceDelegate {
+  finishTask(
+    documentUri: string,
+    graphs: Graph[],
+    diagnostics: Map<string, vscode.Diagnostic[]>,
+  ): void;
+}
+
+class AnalysisTask {
+  private worker: Worker = new Worker(join(__dirname, "./Worker.js"));
+
+  constructor(
+    private documentUri: string,
+    public programs: Program[],
+    private delegate: AnalysisServiceDelegate,
+  ) {
+    this.worker.on(
+      "message",
+      (data: EngineProcessingResult | WorkerLoggerMessage) => {
+        if (data instanceof EngineProcessingResult) {
```

Comment:
I doubt that the class is still identifiable after the postMessage (which is documented as using structured clone algorithm).
That's also one of the reason why I suggested exporting a type and not class.
```
> class X{}
undefined
> var x = new X()
undefined
> x instanceof X
true
> structuredClone(x) instanceof X
false
>
```